sycl/include/sycl/builtins.hpp

```diff
@@ -170,6 +187,100 @@ inline __SYCL_ALWAYS_INLINE
 
 #undef __SYCL_MATH_FUNCTION_2_OVERLOAD_IMPL
 
+#define __SYCL_MATH_FUNCTION_2_SGENFLOAT_Y_OVERLOAD(NAME)                      \
+  template <typename T, size_t N>                                              \
+  inline __SYCL_ALWAYS_INLINE                                                  \
+      std::enable_if_t<detail::is_sgenfloat<T>::value, marray<T, N>>           \
+      NAME(marray<T, N> x, T y) __NOEXC {                                      \
+    marray<T, N> res;                                                          \
+    sycl::vec<T, 2> y_vec{y, y};                                               \
+    for (size_t i = 0; i < N / 2; i++) {                                       \
+      auto partial_res = __sycl_std::__invoke_##NAME<vec<T, 2>>(               \
+          detail::to_vec2(x, i * 2), y_vec);                                   \
+      std::memcpy(&res[i * 2], &partial_res, sizeof(vec<T, 2>));               \
+    }                                                                          \
+    if (N % 2) {                                                               \
+      res[N - 1] = __sycl_std::__invoke_##NAME<T>(x[N - 1], y_vec[0]);         \
+    }                                                                          \
+    return res;                                                                \
+  }
+
+__SYCL_MATH_FUNCTION_2_SGENFLOAT_Y_OVERLOAD(fmax)
+    __SYCL_MATH_FUNCTION_2_SGENFLOAT_Y_OVERLOAD(fmin)
+
+#undef __SYCL_MATH_FUNCTION_2_SGENFLOAT_Y_OVERLOAD
+
+        template <typename T, size_t N>
+        inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<detail::is_sgenfloat<T>::value, marray<T, N>> ldexp(
+        marray<T, N> x, marray<int, N> k) __NOEXC {
+  marray<T, N> res;
+  for (size_t i = 0; i < N; i++) {
+    res[i] = __sycl_std::__invoke_ldexp<T>(x[i], k[i]);
+  }
+  return res;
+}
+
+#undef __SYCL_MATH_FUNCTION_2_GENINT_K_OVERLOAD_IMPL
+
+template <typename T, size_t N>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<detail::is_sgenfloat<T>::value, marray<T, N>>
+    ldexp(marray<T, N> x, int k) __NOEXC {
+  marray<T, N> res;
+  for (size_t i = 0; i < N; i++) {
+    res[i] = __sycl_std::__invoke_ldexp<T>(x[i], k);
+  }
+  return res;
+}
+
+#undef __SYCL_MATH_FUNCTION_2_INT_K_OVERLOAD_IMPL
+
+#define __SYCL_MATH_FUNCTION_2_GENINT_Y_OVERLOAD_IMPL(NAME)                    \
+  marray<T, N> res;                                                            \
+  for (size_t i = 0; i < N; i++) {                                             \
+    res[i] = __sycl_std::__invoke_##NAME<T>(x[i], y[i]);                       \
+  }                                                                            \
+  return res;
+
+template <typename T, size_t N>
+inline __SYCL_ALWAYS_INLINE
+    std::enable_if_t<detail::is_sgenfloat<T>::value, marray<T, N>>
+    pown(marray<T, N> x, marray<int, N> y) __NOEXC {
```

Comment:
Looks like we can add one more high-level macro to generalize definitions of `pown`, `rootn` and `ldexp`. Similar to `__SYCL_MATH_FUNCTION_2_OVERLOAD(atan2)`

Reply:
Basically yes, the reason I didn't do this is to stick to code style in this part of the file.